relay-server/src/services/processor.rs

```diff
@@ -1207,96 +1207,90 @@ impl EnvelopeProcessorService {
         Ok(())
     }
 
-    /// Extract metrics from all envelope items.
-    ///
-    /// Caveats:
-    ///  - This functionality is incomplete. At this point, extraction is implemented only for
-    ///    transaction events.
-    fn extract_metrics(
+    /// Extract transaction metrics.
+    fn extract_transaction_metrics(
         &self,
         state: &mut ProcessEnvelopeState<TransactionGroup>,
         sampling_result: &SamplingResult,
     ) -> Result<(), ProcessingError> {
+        if state.event_metrics_extracted {
+            return Ok(());
+        }
+
         // NOTE: This function requires a `metric_extraction` in the project config. Legacy configs
         // will upsert this configuration from transaction and conditional tagging fields, even if
-        // it is not present in the actual project config payload. Once transaction metric
-        // extraction is moved to generic metrics, this can be converted into an early return.
-        let config = match &state.project_state.config.metric_extraction {
-            ErrorBoundary::Ok(ref config) if config.is_enabled() => Some(config),
-            _ => None,
-        };
+        // it is not present in the actual project config payload.
         let global = self.inner.global_config.current();
-        let global_config = match &global.metric_extraction {
-            ErrorBoundary::Ok(global_config) => global_config,
-            #[allow(unused_variables)]
-            ErrorBoundary::Err(e) => {
-                if_processing!(self.inner.config, {
-                    // Config is invalid, but we will try to extract what we can with just the
-                    // project config.
-                    relay_log::error!("Failed to parse global extraction config {e}");
-                    MetricExtractionGroups::EMPTY
-                } else {
-                    // If there's an error with global metrics extraction, it is safe to assume that this
-                    // Relay instance is not up-to-date, and we should skip extraction.
-                    relay_log::debug!("Failed to parse global extraction config {e}");
-                    return Ok(());
-                })
-            }
+        let generic_config = {
+            let config = match &state.project_state.config.metric_extraction {
+                ErrorBoundary::Ok(ref config) if config.is_supported() => config,
+                _ => return Ok(()),
```

Comment:
Generic project config is no longer optional. `event_metrics_extraced` will only be set if both configs are present, allowing us to bump the two extraction versions individually in the future.